### PR TITLE
Adding Brokerset Overrides

### PR DESCRIPTION
**Pull request by NavalOrange**
Adding a brokersetOverrides option that will change the brokerset list immediately after reading from the list to take all the instances of a set of brokers and move them to a new set of brokers on every brokerset. This means that you can specify a few brokers, use the override to move all the topics off of those specific brokers when you want to make broker changes, and then remove the override file and allow all the topics to go back to their original broker names.

An empty brokersetOverride.json file is acceptable, however any other option where the file isn't empty and doesn't contain all 4 of the required brokersets will result in a failure to parse brokersets. This is by design. The eventually that I'm guarding against is if someone moves topics off of brokers, then changes the override file to something broken, I don't want to assume the user wants to go back to the pre-override brokers. 
tl;dr: Anything except an override success and no override should result in a brokerset parse failure which results in no change in brokers

### CODE REVIEW

File: orion-server/src/test/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensorTest.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class KafkaClusterInfoSensorTest {
```
Comment (vahidhashemian) [inline orion-server/src/test/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensorTest.java:18]: Is there a test that covers the end to end broker replacement scenario? I.e.
1. There is a given brokerset that is affected by a broker-range replacement
2. Override brokerset is created
3. Replacement is complete
4. Override brokerset is removed
5. Before and after brokersets (ranges) are the same.

File: orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java [hunk 2]
```diff
@@ -79,20 +84,38 @@ protected void loadAndSetTopicAssignments(File infoDir,
...
+    File overrideFile = new File(infoDir, "brokersetOverrides.json");
     try {
-      List<Brokerset> brokersets = loadBrokersetsFromFile(file);
+      List<Brokerset> brokersets = loadBrokersetsFromFile(file, overrideFile);
       for (Brokerset brokerset : brokersets) {
         brokersetMap.put(brokerset.getBrokersetAlias(), brokerset);
       }
+    } catch (MalformedJsonException e) {
+      logger.log(Level.SEVERE,
+              "Failed to parse brokerset override file for cluster:" + cluster.getClusterId(), e);
```
[the remark this answers is not in the document]
Comment (NavalOrange) [reply orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java:95]: Good point, I didn't really like how there were 2 exception types here anyway. I think what I'll do is I'll catch these errors in the loadBrokersetsFromFile method and wrap them in another single type of exception that will provide more context

File: orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java [hunk 3]
```diff
@@ -79,20 +84,38 @@ protected void loadAndSetTopicAssignments(File infoDir,
...
+    return loadBrokersetsFromFile(file, null);
+  }
+
+  public static List<Brokerset> loadBrokersetsFromFile(File file, @Nullable File overrides) throws IOException {
+    List<Brokerset> unsanitized = jsonfileToObject(file, new TypeReference<List<Brokerset>>(){});
+    if (overrides == null || !overrides.exists() || overrides.length() == 0) {
+      return unsanitized;
+    }
+    List<Brokerset> sanitized;
+    BrokersetRangeMap map = loadBrokersetOverrides(overrides);
```
[the remark this answers is not in the document]
Comment (NavalOrange) [reply orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java:113]: I chose to do so so I could push sanitization down the stack as far as I could, to the actual parse step. This way I don't need to clutter up the cluster info sensor with boring code making sure that the map in and map out is an equal range, and can hide that where the file is parsed.

File: orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java [hunk 4]
```diff
@@ -93,6 +103,15 @@ public void setPartitions(int partitions) {
     this.partitions = partitions;
   }
 
+  public static Brokerset applyBrokerOverrides(Brokerset inBrokerset, BrokersetRangeMap map) {
```
[the remark this answers is not in the document]
Comment (NavalOrange) [reply orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java:106]: I don't remember why I set it up that way, but its a trivial change and a good one as well, if I run into a wall and remember why I did it in the first place I'll be sure to post an update here.

File: orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java [hunk 5]
```diff
@@ -110,8 +129,13 @@ public BrokersetRange() {
     }
 
     public BrokersetRange(int startBrokerIdx, int endBrokerIdx) {
-      this.startBrokerIdx = startBrokerIdx;
-      this.endBrokerIdx = endBrokerIdx;
+      if (startBrokerIdx > endBrokerIdx) {
```
[the remark this answers is not in the document]
Comment (NavalOrange) [reply orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java:132]: The case I'm guarding against is the user without thinking sets up a file that has `startBrokerIdx = 10` and `endBrokerIdx = 5`, all this is doing is saying that the range of 10-5 is still valid, putting it in a form that won't get us in trouble later and keeping that case in the constructor. I suppose another question would be `is the range of 10-5 a valid range?` Or `is there any case where the range of 10-5 is preferable to 5-10` I think the answer this guard implies for those 2 questions is yes, and no

File: orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java [hunk 6]
```diff
@@ -155,5 +179,114 @@ public void setEndBrokerIdx(int endBrokerIdx) {
       this.endBrokerIdx = endBrokerIdx;
     }
 
+    public Boolean overlaps(BrokersetRange other) {
+      return  !((this.startBrokerIdx < other.startBrokerIdx && this.endBrokerIdx < other.startBrokerIdx) ||
+              (this.startBrokerIdx > other.endBrokerIdx && this.endBrokerIdx > other.endBrokerIdx));
+    }
```
[the remark this answers is not in the document]
Comment (NavalOrange) [reply orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java:185]: Wisdom : D. Thanks Vahid, I wrote this before I wrote the guard in the constructor and didn't go back to change it.

File: orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java [hunk 7]
```diff
@@ -155,5 +179,114 @@ public void setEndBrokerIdx(int endBrokerIdx) {
...
+      BrokersetRange range = (BrokersetRange) o;
+      return startBrokerIdx == range.startBrokerIdx && endBrokerIdx == range.endBrokerIdx;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(startBrokerIdx, endBrokerIdx);
+    }
+
+    public static List<BrokersetRange> applyMapping(BrokersetRange range, BrokersetRangeMap map) {
```
[the remark this answers is not in the document]
Comment (NavalOrange) [reply orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java:203]: Same as above, I think you're right

File: orion-server/src/test/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensorTest.java (hunk 1 above)
Comment (NavalOrange) [reply orion-server/src/test/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensorTest.java:18]: Hmm, not a bad idea, but I don't know how useful it would be here, as each of those operations are idempotent. I mean the way I can think to set up a test like that here would be to have 
`List<Brokerset> result = KafkaClusterInfoSensor.loadBrokersetsFromFile(brokersetFile, overridesFile);`
then 
`List<Brokerset> result2 = KafkaClusterInfoSensor.loadBrokersetsFromFile(brokersetFile);`
then 
`List<Brokerset> result3 = KafkaClusterInfoSensor.loadBrokersetsFromFile(brokersetFile, overridesFile);`
Which the test is going to look at and be perfectly happy to tell me that the two identical calls to an idempotent method are equal. The only thing that would really be testing would be the file system operations at that point (Which if I was going to make these tests better would be the first things I would remove)
That test would be better for a big scary integration test, which at this point I am not prepared to get into.

Comment (NavalOrange) [reply orion-server/src/test/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensorTest.java:18]: And as for getting rid of the tests, if I was told to do that I would have to MIGHTILY resist the urge to convert this test to Kotlin, because that would be another conversation that I don't want to have right now.

File: orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (NavalOrange) [reply orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java:113]: Ah sorry I think I wasn't reading your question properly and started writing a big diatribe about why to have a custom deserializer.

No real reason, just a stylistic decision, more explicit method call for marginally more readable code and cheap as free as I think the compiler would inline it. I can go back

File: orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (NavalOrange) [reply orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java:132]: no and no it is!

File: orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java [hunk 8]
```diff
@@ -120,7 +139,10 @@ protected void loadAndSetConsumerInfo(File infoDir, KafkaCluster cluster) {
 
   public static <T> T jsonfileToObject(File jsonFile, TypeReference<T> typeReference ) throws IOException {
     ObjectMapper mapper = new ObjectMapper();
-    return mapper.readValue(jsonFile, typeReference);
+    try {
+      return mapper.readValue(jsonFile, typeReference);
+    } catch (Exception e){
+      throw new IOException("Problem found when reading " + jsonFile, e);
```
Comment (NavalOrange) [inline orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java:145]: Added the additional context for errors here.

File: orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (NavalOrange) [reply orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java:93]: I added the additional context in the jsonfileToObject method. https://github.com/pinterest/orion/pull/52/commits/e5a966385e0039cdc377f3812c9265518ae7e242#r819053053

[the remark this answers is not in the document]
Comment (NavalOrange) [reply orion-server/src/main/java/com/pinterest/orion/core/automation/sensor/kafka/KafkaClusterInfoSensor.java:93]: Oh geez, I missed a spot in the string there. I'll post another quick update

File: orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java [hunk 9]
```diff
@@ -101,15 +122,17 @@ public String toString() {
...
     private int startBrokerIdx;
     private int endBrokerIdx;
-    
-    public BrokersetRange() {
-    }
 
     public BrokersetRange(int startBrokerIdx, int endBrokerIdx) {
+      if (startBrokerIdx > endBrokerIdx) {
+        throw new IllegalArgumentException("Brokerset Ranges must come in pairs where 'startBrokerIdx' <= 'endBrokerIdx'\n" +
+                "There is a broker set that includes range startBrokerIdx: " + startBrokerIdx + " and endBrokerIdx: " + endBrokerIdx);
```
Comment (vahidhashemian) [inline orion-server/src/main/java/com/pinterest/orion/core/kafka/Brokerset.java:134]: What does this second sentence mean in the exception message?

**Review by vahidhashemian — APPROVED** [review]
Thanks for the great addition. I left one minor comment.
